python/ray/tests/test_runtime_env_failure.py

```diff
@@ -0,0 +1,155 @@
+        },
+    ):
+        print("RAY_RUNTIME_ENV_FAIL_DOWNLOAD_FOR_TESTING enabled.")
+        yield
+
+
+class TestRuntimeEnvFailure:
+    @pytest.mark.parametrize("plugin", ["working_dir", "py_modules"])
+    def test_fail_upload(self, tmpdir, monkeypatch, start_cluster, plugin):
+        """Simulate failing to upload the working_dir to the GCS.
+
+        Test that we raise an exception and don't hang.
+        """
+        monkeypatch.setenv(RAY_RUNTIME_ENV_FAIL_UPLOAD_FOR_TESTING_ENV_VAR, "1")
+        _, address = start_cluster
+        if plugin == "working_dir":
+            runtime_env = {"working_dir": str(tmpdir)}
+        else:
+            runtime_env = {"py_modules": [str(tmpdir)]}
+
+        with pytest.raises(Exception):
+            ray.init(address, runtime_env=runtime_env)
+
+    def test_fail_runtime_env_download(
+        self, tmpdir, monkeypatch, fail_download, start_cluster
+    ):
+        """Simulate failing to download the working_dir from the GCS.
+
+        Test that we raise an exception and don't hang.
+        """
+        _, address = start_cluster
+
+        def init_ray():
+            ray.init(address, runtime_env={"working_dir": str(tmpdir)})
+
+        if using_ray_client(address):
+            # Fails at ray.init() because the working_dir is downloaded for the
+            # Ray Client server.
+            with pytest.raises(Exception):
+                init_ray()
+        else:
+            init_ray()
+            # TODO(architkulkarni): After #25972 is resolved, we should raise an
+            # exception in ray.init().  Until then, we need to `ray.get` a task
+            # to raise the exception.
+
+            @ray.remote
+            def f():
+                pass
+
+            with pytest.raises(Exception):
+                ray.get(f.remote())
+
+    def test_eager_install_fail(self, tmpdir, monkeypatch, start_cluster):
+        """Simulate failing to install a runtime_env in ray.init().
+
+        By default eager_install is set to true.  We should make sure
+        the driver fails to start if the eager_install fails.
+        """
+        _, address = start_cluster
+
+        def init_ray():
+            # Simulate failure using a nonexistent `pip` package.  This will pass
+            # validation but fail during installation.
+            ray.init(address, runtime_env={"pip": ["ray-nonexistent-pkg"]})
+
+        if using_ray_client(address):
+            # Fails at ray.init() because the `pip` package is downloaded for the
+            # Ray Client server.
+            with pytest.raises(Exception):
+                init_ray()
+        else:
+            init_ray()
+
+            # TODO(architkulkarni): After #25972 is resolved, we should raise an
+            # exception in ray.init().  Until then, we need to `ray.get` a task
+            # to raise the exception.
+            @ray.remote
+            def f():
+                pass
+
+            with pytest.raises(Exception):
```

Comment:
I think this is suppose to be `RuntimeEnvSetupError`. Can you catch the precise exceptions here?